### PR TITLE
[TF FE] Support complex tensors for OnesLike operation

### LINKED ISSUE #22953: [Good First Issue][TF FE]: Support complex tensors for OnesLike operation
### Context

OpenVINO component responsible for support of TensorFlow models is called as TensorFlow Frontend (TF FE). TF FE converts a model represented in [TensorFlow opset](https://www.tensorflow.org/api_docs/python/tf/raw_ops) to a model in [OpenVINO opset](https://docs.openvino.ai/2023.1/openvino_docs_ops_opset.html).
Some audio models use tensors of complex type. Complex type tensor is a tensor that has elements of complex type. For example, 1D tensor with three elements `x = [1+2j, 2, -2j]`.

For supporting [OnesLike](https://www.tensorflow.org/api_docs/python/tf/raw_ops/OnesLike) operation on complex type tensor, you need to extend the corresponding [loader for OnesLike](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/tensorflow_common/src/op/ones_like.cpp).

### What needs to be done?

The existing [loader for OnesLike](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/tensorflow_common/src/op/ones_like.cpp) needs to be extended by propagating `ComplexTypeMark` from input to output and to represent output complex type tensor as a floating-point type tensor with auxiliary dimension that concatenates real and imaginary parts of complex tensor.
To validate the extension, the corresponding layer test needs to be updated with complex tensor cases.

Here is an example of how to extend `Reshape` loader to support complex type tensors:

```cpp
OutputVector translate_reshape_op(const NodeContext& node) {
    default_op_checks(node, 2, {"Reshape"}, true);
    auto tensor = node.get_input(0);
    auto complex_type_mark = as_type_ptr<ComplexTypeMark>(tensor.get_node_shared_ptr());
    auto shape = node.get_input(1);
    if (complex_type_mark) {
        element::Type complex_part_type = complex_type_mark->get_complex_part_type();
        tensor = complex_type_mark->input_value(0);

        OutputVector concat_inputs;
        concat_inputs.push_back(shape);
        concat_inputs.push_back(make_shared<v0::Constant>(shape.get_element_type(), Shape{1}, 2));

        auto concat = make_shared<v0::Concat>(concat_inputs, 0);
        auto reshape = make_shared<v1::Reshape>(tensor, concat, false);
        set_node_name(node.get_name(), reshape);
        auto complex_reshape = make_shared<ComplexTypeMark>(reshape, complex_part_type);
        return {complex_reshape->output(0)};
    }

    auto reshape = make_shared<v1::Reshape>(tensor, shape, false);
    set_node_name(node.get_name(), reshape);
    return {reshape};
}
```

Since OpenVINO does not have native support of complex tensors, we handle complex type in intermediate layers by representing them as a floating-point type with additional dimension (specially created) to store real and imaginary parts of the original complex tensor so slicing by the last dimension will give either real or imaginary parts: `x[...,0]` - real and `x[...,1]` - imaginary parts.

On the first step, we update `default_op_checks` with `true` flag to indicate that loader for `Reshape` operation now handles complex tensors:

```
default_op_checks(node, 2, {"Reshape"}, true);
```

Secondly, we check if complex type mark exists by anticipated inputs. This mark indicates that input tensor of complex type:

```
auto complex_type_mark = as_type_ptr<ComplexTypeMark>(tensor.get_node_shared_ptr());
```

Thirdly, we retrieve a floating-point tensor (with additional dimension to store real and imaginary parts) simulating complex tensor:

```
tensor = complex_type_mark->input_value(0);
```

After that, we implement conversion for `Reshape` for this particular case. Since a floating-point tensor simulating complex tensor has additional dimension equal to 2,
we update input target shape by appending `2` value and perform reshape on a floating-point tensor simulating complex tensor.

Finally, since `Reshape` should produce complex tensor by output we insert a new mark `ComplexTypeMark` into the output.

To validate support of complex tensors for `Reshape`, the new layer test [TestComplexReshape](https://github.com/openvinotoolkit/openvino/blob/master/tests/layer_tests/tensorflow_tests/test_tf_Reshape.py#L53) was added.

Example how to run the layer test:

```
export TEST_DEVICE=CPU
cd openvino/tests/layer_tests/tensorflow_tests
pytest test_tf_Reshape.py
```

### Example Pull Requests

- Extensions for `Shape`, `Mul`, `Reshape` in https://github.com/openvinotoolkit/openvino/pull/21477
- Extension for `Roll` operation in https://github.com/openvinotoolkit/openvino/pull/20860

### Resources

- [What is OpenVINO?](https://github.com/openvinotoolkit/openvino#what-is-openvino-toolkit)
- [How to Build OpenVINO](https://github.com/openvinotoolkit/openvino/blob/master/docs/dev/build.md)
- [Developer documentation for TensorFlow Frontend](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/tensorflow)
- [Intel DevHub Discord channel](https://discord.gg/7pVRxUwdWG) - engage in discussions, ask questions and talk to OpenVINO developers

### Contact points

- @openvinotoolkit/openvino-tf-frontend-maintainers
- rkazants in Discord

### Ticket

_No response_

### PR DESCRIPTION
**Pull request by MonalSD**
### Details:
 - *Adding support for OnesLike operation on complex type tensor*
- Closes #22953

### CODE REVIEW

**Comment by MonalSD** [comment]
Hi @rkazants, I have raised a PR for issue #22953 please take a look

File: src/frontends/tensorflow_common/src/op/ones_like.cpp [hunk 1]
```diff
@@ -21,6 +21,11 @@ namespace op {
 OutputVector translate_ones_like_op(const NodeContext& node) {
     default_op_checks(node, 1, {"OnesLike"});
     auto x = node.get_input(0);
+    auto complex_type_mark_x = as_type_ptr<ComplexTypeMark>(x.get_node_shared_ptr());
+    if (complex_type_mark_x) {
+        x = complex_type_mark_x->input_value(0);
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/ones_like.cpp:26]: Not correct implementation:)
That is because we receive ones in imaginary part but imag part should consists of zeros. Please implement separate logic for complex type.

image

**Review by rkazants — CHANGES_REQUESTED** [review]
Need to re-implement.

File: src/frontends/tensorflow_common/src/op/ones_like.cpp (hunk 1 above)
Comment (MonalSD) [reply src/frontends/tensorflow_common/src/op/ones_like.cpp:26]: @rkazants thanks for your input. I'll amend the changes

File: src/frontends/tensorflow_common/src/op/ones_like.cpp [hunk 2]
```diff
@@ -19,27 +19,56 @@ namespace tensorflow {
...
+        shape_of = make_shared<v0::Concat>(OutputVector{one_dim, shape_of}, 0);
+
+        // create a tensor of zeros of shape with extra dimension
+        Output<Node> ones_like = make_shared<v3::Broadcast>(one_const, shape_of);
+        // remove extra dimension by squeezing
+        auto zero_dim_ind = make_shared<v0::Constant>(element::i32, Shape{1}, 0);
+        ones_like = make_shared<v0::Squeeze>(ones_like, zero_dim_ind);
+        set_node_name(node.get_name(), ones_like.get_node_shared_ptr());
+        return {ones_like};
+    }
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/ones_like.cpp:70]: please remove `else` branch and put code outside as it was. It will be equivalent and simpler to review non-complex type case to make sure no changes for non-complex type:)

**Comment by MonalSD** [comment]
@rkazants please take a look at the addressed changes.

**Comment by MonalSD** [comment]
@rkazants please review the changes and give your input. Thankyou

File: src/frontends/tensorflow_common/src/op/ones_like.cpp [hunk 3]
```diff
@@ -19,8 +21,29 @@ namespace tensorflow {
...
+        Output<Node> shape_of_real = make_shared<v3::ShapeOf>(x_real, element::i32);
+
+        auto one_const = create_same_type_const_scalar<int32_t>(x_real, 1);
+        Output<Node> ones_like = make_shared<v3::Broadcast>(one_const, shape_of_real);
+
+        auto zero_const = create_same_type_const_scalar<int32_t>(x_real, 0);
+        Output<Node> zeros_like = make_shared<v3::Broadcast>(zero_const, shape_of_real);
+        auto result = make_shared<v0::Concat>(OutputVector{ones_like, zeros_like}, -1);
+        set_node_name(node.get_name(), result);
+
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/ones_like.cpp:41]:
```suggestion
```

**Comment by rkazants** [comment]
build_jenkins

**Review by rkazants — APPROVED** [review]

File: tests/layer_tests/tensorflow_tests/test_tf_OnesLike.py [hunk 4]
```diff
@@ -43,3 +43,47 @@ def test_ones_like(self, params, ie_device, precision, ir_version, temp_dir,
...
+            tf.compat.v1.global_variables_initializer()
+            tf_net = sess.graph_def
+
+        return tf_net, None
+
+    test_data_basic = [
+        dict(x_shape=[], x_type=np.float32),
+        dict(x_shape=[2], x_type=np.int32),
+        dict(x_shape=[2, 3, 4], x_type=np.float32),
+        dict(x_shape=[1, 4, 3, 1], x_type=np.int32),
```
Comment (rkazants) [inline tests/layer_tests/tensorflow_tests/test_tf_OnesLike.py:79]:
```suggestion
        dict(x_shape=[], x_type=np.float32),
        dict(x_shape=[2], x_type=np.float32),
        dict(x_shape=[2, 3, 4], x_type=np.float32),
        dict(x_shape=[1, 4, 3, 1], x_type=np.float32),
```

**Comment by rkazants** [comment]
build_jenkins

File: src/frontends/tensorflow_common/src/op/ones_like.cpp [hunk 5]
```diff
@@ -32,14 +54,9 @@ OutputVector translate_ones_like_op(const NodeContext& node) {
...
     // create a tensor of zeros of shape with extra dimension
     Output<Node> ones_like = make_shared<v3::Broadcast>(one_const, shape_of);
-    // remove extra dimension by squeezing
-    auto zero_dim_ind = make_shared<v0::Constant>(element::i32, Shape{1}, 0);
-    ones_like = make_shared<v0::Squeeze>(ones_like, zero_dim_ind);
-
     set_node_name(node.get_name(), ones_like.get_node_shared_ptr());
     return {ones_like};
 }
-
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/ones_like.cpp:42]: please revert these changes

**Review by rkazants — COMMENTED** [review]
undesired changes

**Review by rkazants — CHANGES_REQUESTED** [review]
please revert undesired changes

**Comment by MonalSD** [comment]
@rkazants reverted changes please run the build.

File: tests/layer_tests/tensorflow_tests/test_tf_OnesLike.py [hunk 6]
```diff
@@ -43,3 +43,47 @@ def test_ones_like(self, params, ie_device, precision, ir_version, temp_dir,
         self._test(*self.create_ones_like_net(**params),
                    ie_device, precision, ir_version, temp_dir=temp_dir,
                    use_legacy_frontend=use_legacy_frontend)
+
```
Comment (rkazants) [inline tests/layer_tests/tensorflow_tests/test_tf_OnesLike.py:46]:
```suggestion


```

**Comment by rkazants** [comment]
build_jenkins

**Review by rkazants — APPROVED** [review]
